### PR TITLE
Refactor Logpipeline api for secret refs

### PR DESCRIPTION
**Pull request by rakesh-garimella**
**Description**

Changes proposed in this pull request:

- Update the CRD as per latest specification
- If the secret referenced is not present then the pipeline stays in pending. While creating such pipeline user is presented with warning message.
- Included sync period to check watch all secrets periodically.

**Related issue(s)**

https://github.com/kyma-project/kyma/issues/14558

### CODE REVIEW

File: components/telemetry-operator/internal/sync/sync.go [hunk 1]
```diff
@@ -257,56 +258,100 @@ func (s *LogPipelineSyncer) syncFilesConfigMap(ctx context.Context, logPipeline
...
+	newSecret.Data = make(map[string][]byte)
+
+	var logPipelines telemetryv1alpha1.LogPipelineList
+	err = s.List(ctx, &logPipelines)
+	if err != nil {
+		return false, err
+	}
+
+	if logPipeline.DeletionTimestamp != nil {
+		controllerutil.RemoveFinalizer(logPipeline, secretRefsFinalizer)
```
Comment (chrkl) [inline components/telemetry-operator/internal/sync/sync.go:296]: I think we don't need that finalizer anymore since the secret is regenerated on any change. Also with your change it is not added anymore.

File: components/telemetry-operator/internal/sync/sync.go [hunk 2]
```diff
@@ -257,56 +258,100 @@ func (s *LogPipelineSyncer) syncFilesConfigMap(ctx context.Context, logPipeline
...
+	}
+
+	for _, l := range logPipelines.Items {
+		for _, varRef := range l.Spec.Variables {
```
Comment (chrkl) [inline components/telemetry-operator/internal/sync/sync.go:300]: We should skip LogPipelines that have a DeletionTimestamp. Would probably not make a big difference since they will finally disappear and not be added again in the next reconciliation round, but skipping them would immediately result in the desired state.

**Review by chrkl — CHANGES_REQUESTED** [review]
Please also give more details in the PR description

File: components/telemetry-operator/internal/sync/sync.go (hunk 2 above)
Comment (rakesh-garimella) [reply components/telemetry-operator/internal/sync/sync.go:300]: I fetch the current list of logpipelines and here we would not have the deleting logpipeline. So thats why have moved the logic related to deleting log pipeline outside.

File: components/telemetry-operator/main.go [hunk 3]
```diff
@@ -132,8 +133,10 @@ func main() {
 		setupLog.Error(err, "Invalid flag provided")
 		os.Exit(1)
 	}
+	var syncPeriod = 1 * time.Minute
```
[the remark this answers is not in the document]
Comment (rakesh-garimella) [reply components/telemetry-operator/main.go:136]: will do

File: components/telemetry-operator/internal/sync/sync.go (hunk 1 above)
Comment (rakesh-garimella) [reply components/telemetry-operator/internal/sync/sync.go:296]: removed.

File: components/telemetry-operator/internal/sync/sync.go [hunk 4]
```diff
@@ -257,56 +258,100 @@ func (s *LogPipelineSyncer) syncFilesConfigMap(ctx context.Context, logPipeline
 	return changed, nil
 }
 
-// Copy referenced secrets to global Fluent Bit environment secret.
-func (s *LogPipelineSyncer) syncSecretRefs(ctx context.Context, logPipeline *telemetryv1alpha1.LogPipeline) (bool, error) {
+func isSecretRef(fromType telemetryv1alpha1.ValueFromType) bool {
```
[the remark this answers is not in the document]
Comment (rakesh-garimella) [reply components/telemetry-operator/internal/sync/sync.go]: done.

File: components/telemetry-operator/internal/fluentbit/variable_validator.go [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+	for _, variable := range logPipeline.Spec.Variables {
+		err := v.validateSecretKeysExist(context, variable.ValueFrom.SecretKey)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (v *variablesValidator) validateSecretKeysExist(ctx context.Context, secretKeyRef telemetryv1alpha1.SecretKeyRef) error {
```
[the remark this answers is not in the document]
Comment (rakesh-garimella) [reply components/telemetry-operator/internal/fluentbit/variable_validator.go]: done.

File: components/telemetry-operator/internal/fluentbit/variable_validator.go [hunk 6]
```diff
@@ -0,0 +1,63 @@
+package fluentbit
```
Comment (chrkl) [inline components/telemetry-operator/internal/fluentbit/variable_validator.go:1]: This is not really related to fluentbit. How about moving all the validators to an own `validation` package?

File: components/telemetry-operator/internal/secret/secret_handling.go [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+	"fmt"
+
+	telemetryv1alpha1 "github.com/kyma-project/kyma/components/telemetry-operator/api/v1alpha1"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	logf "sigs.k8s.io/controller-runtime/pkg/log"
+)
+
+type SecretValidator struct {
```
Comment (chrkl) [inline components/telemetry-operator/internal/secret/secret_handling.go:15]: Since this does not only validate secrets, let's find a better name. How about `SecretHelper`?

File: components/telemetry-operator/internal/fluentbit/variable_validator_test.go [hunk 8]
```diff
@@ -0,0 +1,99 @@
...
+	"testing"
+
+	telemetryv1alpha1 "github.com/kyma-project/kyma/components/telemetry-operator/api/v1alpha1"
+	"github.com/kyma-project/kyma/components/telemetry-operator/internal/sync/mocks"
+	"github.com/stretchr/testify/require"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+func TestValidateSecretRefs(t *testing.T) {
+
```
Comment (chrkl) [inline components/telemetry-operator/internal/fluentbit/variable_validator_test.go:14]: remove

File: components/telemetry-operator/internal/webhook/logpipeline_webhook.go [hunk 9]
```diff
@@ -101,6 +106,17 @@ func (v *LogPipelineValidator) Handle(ctx context.Context, req admission.Request
 		}
 	}
 
+	secretsFound := v.validateSecrets(ctx, logPipeline)
+	if !secretsFound {
+		warnMsg := "One or more secrets do not exist, the log pipeline creation would not be completed"
```
Comment (chrkl) [inline components/telemetry-operator/internal/webhook/logpipeline_webhook.go:111]: Let's be more precise about the behaviour in the warn message. I think it might be unclear if the user has to fix this, or it will resolve itself. Better write that applying the configuration will be pending until all secrets are available.

**Review by chrkl — APPROVED** [review]

**Review by chrkl — APPROVED** [review]
